20.11 fb field export import automated test

labkey-chrisj opened the pull request:
#### Rationale
this adds test coverage for the new field export/import functionality in the domain designer

#### Related Pull Requests
n/a

#### Changes
update domainFieldRow page class to download fields exported as json
update webDriverWrapper to cause firefox to automatically download application/json content

labkey-chrisj committed:
DomainDesigner field import/export testing

labkey-chrisj committed:
DomainDesigner field import/export testing

cnathe commented on src/org/labkey/test/components/domain/DomainFormPanel.java:
```diff
@@ -288,6 +306,9 @@ public void click()
         };
 
+        protected WebElement exportFieldsButton = Locator.tagWithClass("div", "domain-toolbar-export-btn")
+                .findWhenNeeded(this);
```
The add field button finder uses `.findWhenNeeded(DomainFormPanel.this)` so that it has the given panel in scope when looking for the button. Is that needed here as well for the case when there are multiple domain form panels on the page?

labkey-chrisj replied:
The addFieldButton finder is enclosed in an enclosed type so it can override click(); the DomainFormPanel.this notation allows the enclosed class member to up-reference the containing class.  The exportFieldsButton doesn't need to do that, it's a direct member of the current DomainFormPanel instance.

cnathe replied:
Got it. Thanks for explaining.

cnathe commented on src/org/labkey/test/tests/DomainDesignerTest.java:
```diff
@@ -1498,6 +1507,172 @@ public void addUpdateRemoveConditionalFormat() throws Exception
+        List<PropertyDescriptor> createdFields = getFieldsFromExportFile(exportFile);
+
+        PropertyDescriptor stringField = createdFields.stream().filter(a-> a.getName().equals("textField"))
+                .findFirst().orElse(null);
```
This filter to findFirst works, but we should expect a specific order to the fields that are exported and read in from the JSON file. So would it make sense to get the field at a specific index (zero here) since we are using an ArrayList in the getFieldsFromExportFile? This would also then provide a check that the exported fields are in the correct/expected ordering.

labkey-chrisj replied:
I hadn't been enforcing order because field order here is initially established by the CreateDomainCommand, I don't know that the order isn't shuffled along the line while serializing.
I have added code to ensure that field order in the export file matches field order in the UI, also that it is maintained over a round-trip

cnathe commented on src/org/labkey/test/tests/DomainDesignerTest.java:
```diff
@@ -1498,6 +1507,172 @@ public void addUpdateRemoveConditionalFormat() throws Exception
+
+        PropertyDescriptor stringField = createdFields.stream().filter(a-> a.getName().equals("textField"))
+                .findFirst().orElse(null);
+        assertNotNull("expected string field was not exported", stringField);
+        assertThat("description did not export", stringField.getDescription(), is("is texty"));
+        assertThat("label did not export", stringField.getLabel(), is("text field"));
```
If you put the initial FieldDefinitions into variables, you could use them here instead of duplicating the text strings for the label/descriptions/etc.

labkey-chrisj replied:
I've taken that code out, it is redundant.  The validate function covers all of these points for all fields now

labkey-chrisj replied:
Good call.  I've factored them into a list and am using them to validate

cnathe commented on src/org/labkey/test/tests/DomainDesignerTest.java:
```diff
@@ -1498,6 +1507,172 @@ public void addUpdateRemoveConditionalFormat() throws Exception
+            PropertyDescriptor exportedField = createdFields.stream().filter(a-> a.getName().equals(intendedField.getName()))
+                    .findFirst().orElse(null);
+            assertNotNull("expected field ["+intendedField.getName()+"] was not exported", exportedField);
+            verifyExpectedFieldProperties(intendedField, exportedField);
```
I like this "iterate through the dgen.getColumns() and compare with the exportedField" approach. I'm curious why then, the above checks exist on a per field basis? From what I can see the verifyExpectedFieldProperties covers all of the things that those checks before the for loop are doing. Am I missing something?

labkey-chrisj replied:
I wrote those first, then added the iterative/parameterized one after and forgot to remove them

cnathe commented on src/org/labkey/test/tests/DomainDesignerTest.java:
```diff
@@ -1498,6 +1507,172 @@ public void addUpdateRemoveConditionalFormat() throws Exception
+                FieldDefinition.RangeType.LTE, "2020-11-06");
+        TestDataGenerator dgen = new TestDataGenerator(lookupInfo)
+                .withColumns(List.of(
+                        new FieldDefinition("name", FieldDefinition.ColumnType.String),
+                        new FieldDefinition("textField", FieldDefinition.ColumnType.String)
+                                .setDescription("is texty").setLabel("text field"),
+                        new FieldDefinition("booleanField", FieldDefinition.ColumnType.Boolean)
+                                .setLabel("Boolean Field")
+                                .setRequired(true)
+                                .setURL("fake/list-def.url"),
+                        new FieldDefinition("attachmentField", FieldDefinition.ColumnType.Attachment)
+                                .setHidden(true),
+                        new FieldDefinition("decimalField", FieldDefinition.ColumnType.Decimal).
+                                setScale(1234).setFormat("0.####"),
+                        new FieldDefinition("intField", FieldDefinition.ColumnType.Integer)
+                                .setMvEnabled(true)
+                                .setConditionalFormats(List.of(new ConditionalFormat(List.of(gtFive), true, true, false))),
+                        new FieldDefinition("dateTimeField", FieldDefinition.ColumnType.DateAndTime)
+                                .setValidators(List.of(soonValidator))));
```
Looks like this list of field definitions is a duplicate from the other test case. Could this be factored out to a function or variable to be reused in both places?

labkey-chrisj replied:
Yes, but List.of(...) produces an immutable list, and I need to add a 'name' field for sampleTypes, so my refactor produces a mutable list

cnathe commented on src/org/labkey/test/tests/DomainDesignerTest.java:
```diff
@@ -1498,6 +1507,172 @@ public void addUpdateRemoveConditionalFormat() throws Exception
+                            .setConditionalFormats(List.of(new ConditionalFormat(List.of(eqFive), true, true, false))),
+                        new FieldDefinition("dateTimeField", FieldDefinition.ColumnType.DateAndTime)
+                            .setValidators(List.of(soonValidator))));
+        dgen.createList(createDefaultConnection(), "Key");
```
The one "extra" thing for lists is that the exported field definitions JSON has an `isPrimaryKey` property set to true for the key field. Can we be sure to check that in this list definition export case?

labkey-chrisj replied:
thanks, I'll add checks for this.
Unfortunately, that's not a field on the PropertyDescriptor class (tho really, it should be)

cnathe commented on src/org/labkey/test/tests/DomainDesignerTest.java:
```diff
@@ -1498,6 +1507,172 @@ public void addUpdateRemoveConditionalFormat() throws Exception
+        for (PropertyDescriptor intendedField : dgen.getColumns())
+        {
+            String fieldName = intendedField.getName();
+            if (!fieldName.equals("name"))  // don't expect name col to export from sampletype, it is special
```
The name field for sample type is special in that we add it automatically when the sample type is first saved. The export feature only exports whatever is "in view" for the domain form panel when that button is clicked. So in the sample type case, if you click export before initially saving the sample type, you probably would not get a "name" field, but if you save and then go back into the sample type and then export, you will.

labkey-chrisj replied:
I've updated the test to get the fields from the domain-create response, instead of the ones I sent in.
This means I get back name and key fields in sampleType and list, respectively- and I validate export/import on all of them.

labkey-tchad replied:
That seems dangerous. If our API helper isn't setting some property correctly, it wouldn't be in the create response and we wouldn't end up validating it.

labkey-chrisj replied:
OK, I'll also validate here that everything we send in comes back as well

labkey-chrisj replied:
Doing so shows that the 'name' field isn't on the sampleType when it is created.
It's also not shown on the UI, not exported, cannot be imported, cannot be added via the UI... but the remoteAPI fails if no 'name' field is supplied when creating the sampleType
This has all kinds of code smells and I want this to be a bug, but something tells me it's by design

labkey-tchad commented on src/org/labkey/test/tests/DomainDesignerTest.java:
```diff
@@ -1498,6 +1507,172 @@ public void addUpdateRemoveConditionalFormat() throws Exception
+        // on FieldDefinition, but not gettable on PropertyDescriptor
+    }
+
+    private List<PropertyDescriptor> getFieldsFromExportFile(File exportFile) throws Exception
+    {
+        JSONParser parser = new JSONParser();
+        JSONArray  jsonArray = (JSONArray) parser.parse(new FileReader(exportFile));
+
+        List<PropertyDescriptor> exportFields = new ArrayList<>();
+        for (int i=0; i < jsonArray.size(); i++)
+        {
+            exportFields.add(new PropertyDescriptor((JSONObject) jsonArray.get(i)));
+        }
+
+        return exportFields;
+    }
```
Have this return a `Map<String, PropertyDescriptor>`. The `fields.stream().filter(a-> a.getName().equals("intField"))...` pattern is a bit cumbersome and we have the field names readily available here.
```suggestion
    private Map<String, PropertyDescriptor> getFieldsFromExportFile(File exportFile) throws Exception
    {
        JSONParser parser = new JSONParser();
        JSONArray  jsonArray = (JSONArray) parser.parse(new FileReader(exportFile));

        Map<String, PropertyDescriptor> exportFields = new HashMap<>();
        for (int i=0; i < jsonArray.size(); i++)
        {
            PropertyDescriptor pd = new PropertyDescriptor((JSONObject) jsonArray.get(i));
            exportFields.put(pd.getName(), pd);
        }

        return exportFields;
    }
```

labkey-chrisj replied:
OK, I'll do this.
Unfortunately, this breaks my previous approach of ensuring that field ordering survives export/round-trip unless I use a map type that ensures key ordering

labkey-tchad commented on src/org/labkey/test/tests/DomainDesignerTest.java:
```diff
@@ -1498,6 +1507,172 @@ public void addUpdateRemoveConditionalFormat() throws Exception
+                    .filter(a-> a.toJSON().equals(intendedFormat.toJSON())).findFirst().orElse(null);
+            assertNotNull("conditional formats did not export with full fidelity", actualFormat);
+        }
+        // would like to do validators, but this part of the remoteAPI is incomplete; validators are settable
+        // on FieldDefinition, but not gettable on PropertyDescriptor
```
There's no reason we couldn't add a constructor to `FieldDefinition` that takes a `JSONObject` similar to the one in `PropertyDescriptor`.

labkey-chrisj replied:
I want this, but am worried about feature-creep in the context of this story.  Perhaps in another story?

labkey-chrisj committed:
code review feedback:
- verify intended fields vs created fields
- verify created fields to exported and round-tripped ones
- refactor duplicate code into shared functions

cnathe commented on src/org/labkey/test/tests/DomainDesignerTest.java:
```diff
@@ -1498,6 +1509,196 @@ public void addUpdateRemoveConditionalFormat() throws Exception
+
+        domainDesignerPage = DomainDesignerPage.beginAt(this, getProjectName(), "lists", roundTripList);
+        assertThat("expect list created from export to have the same fields",
+                domainDesignerPage.fieldsPanel().fieldNames(), is(uiFields));
```
any reason to also check the field properties here? or do we think that the field names check is sufficient?

cnathe replied:
oh nevermind, that is already being done in testFileImportFieldsToNewSampleType

cnathe reviewed (commented):
Test cases look good to me. I'll defer to Trey for final review/approval.

labkey-chrisj committed:
add test case for import of PK field to sampleType

labkey-tchad reviewed (approved):
Good for now. Need a follow-on story to verify properties that aren't understood by `PropertyDescriptor`, other domain types (e.g. study datasets, assay batch/run/data, VarList), and column types (e.g. Lookup, File, Sample). We should also test some unusual/unsupported scenarios such as importing column lists with incompatible columns for a particular domain (different domains allow File, Attachment, or neither).
